File: ralph-workflow/src/language_detector/scanner.rs
//! File system scanning utilities.
//!
//! Functions for scanning directories and counting file extensions.

/// Maximum directory depth to search for signature files
pub(super) const MAX_SIGNATURE_SEARCH_DEPTH: usize = 6;

/// Check if a directory name should be skipped during scanning.
pub(super) fn should_skip_dir_name(name: &str) -> bool {
    if name.starts_with('.') {
        return true;
    }
    matches!(
        name,
        "node_modules"
            | "target"
            | "dist"
            | "build"
            | "vendor"
            | "__pycache__"
            | "venv"
            | ".venv"
            | "env"
    )
}

pub(super) fn is_test_file_name(
    file_name: &str,
    primary_lang: &str,
    path_components: &[String],
) -> bool {
    match primary_lang {
        "Rust" => {
            file_name == "tests.rs"
                || file_name.ends_with("_test.rs")
                || (std::path::Path::new(file_name)
                    .extension()
                    .is_some_and(|ext| ext.eq_ignore_ascii_case("rs"))
                    && path_components.windows(1).any(|w| w[0] == "tests"))
        }
        "Python" => {
            let has_py_ext = std::path::Path::new(file_name)
                .extension()
                .is_some_and(|ext| ext.eq_ignore_ascii_case("py"));
            (file_name.starts_with("test_") && has_py_ext) || file_name.ends_with("_test.py")
        }
        "JavaScript" | "TypeScript" => {
            file_name.ends_with(".test.js")
                || file_name.ends_with(".spec.js")
                || file_name.ends_with(".test.ts")
                || file_name.ends_with(".spec.ts")
                || file_name.ends_with(".test.tsx")
                || file_name.ends_with(".spec.tsx")
        }
        "Go" => file_name.ends_with("_test.go"),
        "Java" => {
            path_components
                .windows(2)
                .any(|w| w[0] == "src" && w[1] == "test")
                || file_name.ends_with("test.java")
        }
        "Ruby" => file_name.ends_with("_spec.rb") || file_name.ends_with("_test.rb"),
        _ => file_name.contains("test") || file_name.contains("spec"),
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(super) enum SearchResult {
    Found,
    Continue { new_queue: Vec<(PathBuf, usize)> },
    Done,
}

use std::path::PathBuf;

pub(super) fn advance_search(
    queue: &[(PathBuf, usize)],
    file_names: &[(PathBuf, String)],
    _scanned_files: usize,
    primary_lang: &str,
) -> SearchResult {
    let found = file_names.iter().any(|(path, name_lower)| {
        let path_components: Vec<String> = path
            .components()
            .map(|c| c.as_os_str().to_string_lossy().to_lowercase())
            .collect();

        if should_skip_dir_name(name_lower) {
            return false;
        }

        if path.is_dir() {
            matches!(name_lower.as_str(), "tests" | "test" | "spec" | "__tests__")
        } else {
            is_test_file_name(name_lower, primary_lang, &path_components)
        }
    });

    let new_queue: Vec<_> = file_names
        .iter()
        .filter(|(path, name_lower)| {
            path.is_dir()
                && !should_skip_dir_name(name_lower)
                && !matches!(name_lower.as_str(), "tests" | "test" | "spec" | "__tests__")
                && queue
                    .first()
                    .map_or(false, |(_, depth)| *depth < MAX_SIGNATURE_SEARCH_DEPTH)
        })
        .map(|(path, _)| {
            let depth = queue.first().map_or(0, |(_, d)| *d);
            (path.clone(), depth + 1)
        })
        .collect();

    if found {
        SearchResult::Found
    } else if new_queue.is_empty() {
        SearchResult::Done
    } else {
        SearchResult::Continue { new_queue }
    }
}

pub(super) const SIGNATURE_FILE_NAMES: &[&str] = &[
    "cargo.toml",
    "pyproject.toml",
    "requirements.txt",
    "setup.py",
    "pipfile",
    "package.json",
    "package-lock.json",
    "yarn.lock",
    "pnpm-lock.yaml",
    "bun.lockb",
    "bun.lock",
    "gemfile",
    "go.mod",
    "pom.xml",
    "build.gradle",
    "build.gradle.kts",
    "composer.json",
    "mix.exs",
    "pubspec.yaml",
];

pub(super) fn is_signature_file(name: &str) -> bool {
    SIGNATURE_FILE_NAMES.contains(&name)
}

#[derive(Debug, Clone)]
pub(super) struct ScanItem {
    pub path: PathBuf,
    pub name: String,
}

#[derive(Debug, Clone)]
pub(super) enum ScanResult {
    Done {
        matched: Vec<(PathBuf, String)>,
    },
    Continue {
        matched: Vec<(PathBuf, String)>,
        next_items: Vec<ScanItem>,
    },
}

pub(super) fn advance_scan(items: &[ScanItem], depth: usize) -> ScanResult {
    if items.is_empty() {
        return ScanResult::Done {
            matched: Vec::new(),
        };
    }

    let matched: Vec<_> = items
        .iter()
        .filter(|item| !item.path.is_dir() && is_signature_file(&item.name))
        .map(|item| (item.path.clone(), item.name.clone()))
        .collect();

    let dirs: Vec<_> = items
        .iter()
        .filter(|item| {
            item.path.is_dir()
                && !should_skip_dir_name(&item.name)
                && depth < MAX_SIGNATURE_SEARCH_DEPTH
        })
        .cloned()
        .collect();

    if dirs.is_empty() {
        ScanResult::Done { matched }
    } else {
        ScanResult::Continue {
            matched,
            next_items: dirs,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::language_detector::{count_extensions_with_workspace, detect_tests_with_workspace};
    use crate::workspace::MemoryWorkspace;

    #[test]
    fn test_count_extensions_with_workspace() {
        let workspace = MemoryWorkspace::new_test()
            .with_file("src/main.rs", "fn main() {}")
            .with_file("src/lib.rs", "pub mod foo;")
            .with_file("src/foo.rs", "pub fn foo() {}")
            .with_file("Cargo.toml", "[package]");

        let counts = count_extensions_with_workspace(&workspace, std::path::Path::new("")).unwrap();

        assert_eq!(counts.get("rs"), Some(&3));
        assert_eq!(counts.get("toml"), Some(&1));
    }

    #[test]
    fn test_count_extensions_with_workspace_skips_hidden() {
        let workspace = MemoryWorkspace::new_test()
            .with_file("src/main.rs", "fn main() {}")
            .with_file(".git/config", "hidden")
            .with_file(".hidden/file.rs", "hidden");

        let counts = count_extensions_with_workspace(&workspace, std::path::Path::new("")).unwrap();

        // Should only count src/main.rs, not hidden files
        assert_eq!(counts.get("rs"), Some(&1));
    }

    #[test]
    fn test_count_extensions_with_workspace_skips_node_modules() {
        let workspace = MemoryWorkspace::new_test()
            .with_file("src/index.js", "export default {}")
            .with_file("node_modules/lodash/index.js", "module.exports = {}")
            .with_file("node_modules/react/index.js", "module.exports = {}");

        let counts = count_extensions_with_workspace(&workspace, std::path::Path::new("")).unwrap();

        // Should only count src/index.js
        assert_eq!(counts.get("js"), Some(&1));
    }

    #[test]
    fn test_detect_tests_with_workspace_finds_test_dir() {
        let workspace = MemoryWorkspace::new_test()
            .with_file("src/main.rs", "fn main() {}")
            .with_file("tests/integration.rs", "#[test] fn test() {}");

        let has_tests = detect_tests_with_workspace(&workspace, std::path::Path::new(""), "Rust");

        assert!(has_tests);
    }

    #[test]
    fn test_detect_tests_with_workspace_finds_test_files() {
        let workspace = MemoryWorkspace::new_test()
            .with_file("src/main.rs", "fn main() {}")
            .with_file("src/foo_test.rs", "#[test] fn test() {}");

        let has_tests = detect_tests_with_workspace(&workspace, std::path::Path::new(""), "Rust");

        assert!(has_tests);
    }

    #[test]
    fn test_detect_tests_with_workspace_no_tests() {
        let workspace = MemoryWorkspace::new_test()
            .with_file("src/main.rs", "fn main() {}")
            .with_file("src/lib.rs", "pub mod foo;");

        let has_tests = detect_tests_with_workspace(&workspace, std::path::Path::new(""), "Rust");

        assert!(!has_tests);
    }
}

File: ralph-workflow/ralph/language_detector/scanner.py
"""Workspace scanning utilities for language detection."""

from __future__ import annotations

from collections import deque
from pathlib import PurePosixPath
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Iterator

    from ralph.workspace.protocol import Workspace

MAX_FILES_TO_SCAN = 2000
MAX_SIGNATURE_SEARCH_DEPTH = 6
SKIP_DIR_NAMES: set[str] = {
    "node_modules",
    "target",
    "dist",
    "build",
    "vendor",
    "__pycache__",
    "venv",
    ".venv",
    "env",
}
SIGNATURE_FILE_NAMES: set[str] = {
    "cargo.toml",
    "pyproject.toml",
    "requirements.txt",
    "setup.py",
    "pipfile",
    "package.json",
    "package-lock.json",
    "yarn.lock",
    "pnpm-lock.yaml",
    "bun.lockb",
    "bun.lock",
    "gemfile",
    "go.mod",
    "pom.xml",
    "build.gradle",
    "build.gradle.kts",
    "composer.json",
}
TEST_DIRECTORY_NAMES: set[str] = {"tests", "test", "spec", "__tests__"}


def normalize_path(path: str) -> str:
    normalized = str(PurePosixPath(path))
    return "" if normalized in {"", "."} else normalized


def join_path(parent: str, child: str) -> str:
    if not parent:
        return normalize_path(child)
    return normalize_path(str(PurePosixPath(parent) / child))


def should_skip_dir_name(name: str) -> bool:
    lowered = name.lower()
    return lowered.startswith(".") or lowered in SKIP_DIR_NAMES


def iter_files(workspace: Workspace, root: str = "") -> Iterator[str]:
    queue: deque[str] = deque([normalize_path(root)])
    visited: set[str] = set()
    scanned = 0

    while queue and scanned < MAX_FILES_TO_SCAN:
        current = queue.popleft()
        if current in visited:
            continue
        visited.add(current)

        try:
            entries = workspace.list_dir(current)
        except FileNotFoundError:
            continue

        for entry in entries:
            child_path = join_path(current, entry)
            if workspace.is_dir(child_path):
                if not should_skip_dir_name(entry):
                    queue.append(child_path)
                continue
            if workspace.is_file(child_path):
                scanned += 1
                yield child_path
                if scanned >= MAX_FILES_TO_SCAN:
                    break


def count_extensions(workspace: Workspace, root: str = "") -> dict[str, int]:
    counts: dict[str, int] = {}
    for path in iter_files(workspace, root):
        suffix = PurePosixPath(path).suffix
        if not suffix:
            continue
        extension = suffix[1:].lower()
        counts[extension] = counts.get(extension, 0) + 1
    return counts


def collect_signature_files(workspace: Workspace, root: str = "") -> dict[str, list[str]]:
    signatures: dict[str, list[str]] = {}
    queue: deque[tuple[str, int]] = deque([(normalize_path(root), 0)])

    while queue:
        current, depth = queue.popleft()
        try:
            entries = workspace.list_dir(current)
        except FileNotFoundError:
            continue

        for entry in entries:
            entry_lower = entry.lower()
            child_path = join_path(current, entry)
            if workspace.is_dir(child_path):
                if depth < MAX_SIGNATURE_SEARCH_DEPTH and not should_skip_dir_name(entry_lower):
                    queue.append((child_path, depth + 1))
                continue
            if workspace.is_file(child_path) and entry_lower in SIGNATURE_FILE_NAMES:
                signatures.setdefault(entry_lower, []).append(child_path)

    return signatures


def is_test_file_name(file_name: str, primary_language: str, path_components: list[str]) -> bool:
    lower_name = file_name.lower()
    language_checks = {
        "Go": lambda: lower_name.endswith("_test.go"),
        "PHP": lambda: lower_name.endswith("test.php") or lower_name.endswith("spec.php"),
        "Python": lambda: (
            (lower_name.startswith("test_") and lower_name.endswith(".py"))
            or lower_name.endswith("_test.py")
        ),
        "Ruby": lambda: lower_name.endswith("_spec.rb") or lower_name.endswith("_test.rb"),
    }
    if primary_language == "Rust":
        return (
            lower_name == "tests.rs"
            or lower_name.endswith("_test.rs")
            or (lower_name.endswith(".rs") and "tests" in path_components)
        )
    if primary_language in {"JavaScript", "TypeScript"}:
        return any(
            lower_name.endswith(suffix)
            for suffix in (
                ".test.js",
                ".spec.js",
                ".test.ts",
                ".spec.ts",
                ".test.tsx",
                ".spec.tsx",
            )
        )
    if primary_language == "Java":
        return ("src" in path_components and "test" in path_components) or lower_name.endswith(
            "test.java"
        )
    check = language_checks.get(primary_language)
    if check is not None:
        return check()
    return "test" in lower_name or "spec" in lower_name


def detect_tests(workspace: Workspace, root: str = "", primary_language: str = "Unknown") -> bool:
    queue: deque[str] = deque([normalize_path(root)])
    scanned = 0

    while queue and scanned < MAX_FILES_TO_SCAN:
        current = queue.popleft()
        try:
            entries = workspace.list_dir(current)
        except FileNotFoundError:
            continue

        for entry in entries:
            entry_path = join_path(current, entry)
            entry_lower = entry.lower()
            if workspace.is_dir(entry_path):
                if entry_lower in TEST_DIRECTORY_NAMES:
                    return True
                if should_skip_dir_name(entry_lower):
                    continue
                queue.append(entry_path)
                continue
            if workspace.is_file(entry_path):
                scanned += 1
                components = [part.lower() for part in PurePosixPath(entry_path).parts]
                if is_test_file_name(entry, primary_language, components):
                    return True
                if scanned >= MAX_FILES_TO_SCAN:
                    break

    return False
